### DISCUSSION TITLE
Which formats of external subtitles are supported?

### DISCUSSION THREAD

**Discussion by experk** [gerbera/gerbera#1102] [Q&A] [answered]
Hello,
I've been trying to figure this out, but despite subtitles being mentioned multiple time in documentation, I failed to find a simple answer to the subject.

**Comment by KarlStraussberger** [comment]
You can use any format you player supports. Once you registered the pattern in the config resource section it is attached to the media entry in the database and streamed in parallel to the media (sent via UPNP).

**Reply by experk** [reply]
Thank you for swift reply!
Would you mind sharing an example of such config? 
So far I'm not getting any luck with either .srt or .ass, only failure to play a video which was working just fine till I placed a similarly named subtitle file next to it. And to make matters worse debug log only says:

> Failed to send data in request for `/content/media/object_id/59/res_id/0/ext/file.mkv'.

> Failed to send data in request for `/content/media/object_id/63/res_id/0/ext/file.mkv'.

So I don't know in which direction I should dig. I'm running 1.7 from entware

**Comment by KarlStraussberger** [comment, marked as answer]
1.7 is the current HEAD, so you may get an update each day and the version you have, may not be stable. Maybe you can check for a new one or support us by building it on your own and become a tester.

You can still place the file next to the video, but you have to tell gerbera to use it as a subtitle file. The main difference is that now the subtitle file is store as such with the media information so we "know" what to send instead of checking the disk each time. In order to become more flexible we need to configure it:
```
<import >
...
  <resources>
...
    <subtitle>
      <add-file name="%filename%.srt"/>
    </subtitle>
    <resource>
      <add-file name="%filename%.srt"/>
    </resource>
  </resources>
</import>
```

**Reply by experk** [reply]
Thank you for clarification!
I think this is exactly how I have it in my config, yet it didn't work for me. If I'll have more spare time, I'll try to play around with this, but so far I have to failback to minidlna which at least works with .srt.

Also, another thing I've noticed - entware package installer claims to have gerbera 1.6.4, yet once installation is done, I get 1.7. A bit of truncated output from my mipsel router:
```
~ # opkg info gerbera
Package: gerbera
Version: 1.6.4-1
Depends: libc, libssp, librt, libpthread, file, libnpupnp, libsqlite3, libexif, libuuid, libfmt, libffmpeg, libiconv-full, curl, duktape, libffmpegthumbnailer
Status: unknown ok not-installed
Section: multimedia
Architecture: mipsel-3.4
Size: 3173837
Filename: gerbera_1.6.4-1_mipsel-3.4.ipk
Description: Gerbera - UPnP Media Server based on Mediatomb.

~ # opkg install gerbera
Installing gerbera (1.6.4-1) to root...
Downloading http://bin.entware.net/mipselsf-k3.4/gerbera_1.6.4-1_mipsel-3.4.ipk
...
Configuring gerbera.
~ # gerbera
2021-01-08 12:33:29   info: Gerbera UPnP Server version 1.7.0_git - http://gerbera.io/
2021-01-08 12:33:29   info: ===============================================================================
2021-01-08 12:33:29   info: Gerbera is free software, covered by the GNU General Public License version 2
2021-01-08 12:33:29   info: Copyright 2016-2020 Gerbera Contributors.
2021-01-08 12:33:29   info: Gerbera is based on MediaTomb: Copyright 2005-2010 Gena Batsyan, Sergey Bostandzhyan, Leonhard Wimmer.
2021-01-08 12:33:29   info: ===============================================================================
```

**Reply by Freealx** [reply]
I could second the statement for the last current v1.11.0 - the subtitles are still NOT working.
Same subtitle files are perfectly recognized and used from different UPnP/DLNA servers - ReadyMedia (miniDLNA), Mezzmo etc.

**Reply by KarlStraussberger** [reply]
> I could second the statement for the last current v1.11.0 - the subtitles are still NOT working. Same subtitle files are perfectly recognized and used from different UPnP/DLNA servers - ReadyMedia (miniDLNA), Mezzmo etc.

Can you please create an issue describing you situation incl. configuration and clients.

**Reply by KarlStraussberger** [reply]
> I could second the statement for the last current v1.11.0 - the subtitles are still NOT working. Same subtitle files are perfectly recognized and used from different UPnP/DLNA servers - ReadyMedia (miniDLNA), Mezzmo etc.

Just created a PR that allows tweaking mimetype because some clients require the correct mimetype for the subtitle (e.g. Kodi). In order to get it work, you need to set an entry in https://docs.gerbera.io/en/latest/config-import.html#extension-mimetype and optionally map the mimetype to your device's favourite in new `client` subsection

**Reply by Freealx** [reply]
Hi Karl. Thank you for the prompt reply and action!
I'm not really sure what should be additionally configured though...
The mime-types for the external ASCII files subtitles, should be already covered to provide the basic function support - in my case that would be something like:
`<map from="srt" to="text/plain"/>`
Are you saying that there are something else that needs to be done in order to see the subtitles on some specific DLNA players, when using Gerbera?
I mean, something that's already preconfigured on a servers like miniDLNA and Mezzmo?

**Reply by KarlStraussberger** [reply]
If you check out the web ui, you can see what gerbera discovered during import. If that represents what you expect, there are several client config things you may test to get the required UPNP output (mineDLNA does only simple things compared to gerbara) - run with `--debug` to get what we create.

**Reply by Freealx** [reply]
I'm afraid I still cannot understand you, Karl.
In `config/config.xml` I can see that:
```
    <mappings>
      <extension-mimetype ignore-unknown="no">
        <map from="...
                   ... />
      </extension-mimetype>
      <mimetype-upnpclass>
        <map from="...
                   ... />
      </mimetype-upnpclass>
      <mimetype-contenttype>
        <treat mimetype="...
                         ... />
      </mimetype-contenttype>
      <mimetype-dlnatransfermode>
        <map from="application/ogg" to="Streaming" />
        <map from="application/x-srt" to="Background" />
        <map from="audio/*" to="Streaming" />
        <map from="image/*" to="Interactive" />
        <map from="srt" to="Background" />            <---------
        <map from="text/*" to="Background" />
        <map from="video/*" to="Streaming" />
      </mimetype-dlnatransfermode>
      <contenttype-dlnaprofile>
        <map from= ...
                   ... />
      </contenttype-dlnaprofile>
    </mappings>
```
Do we need something else in another mimetype section in order to see the subtitles?
About the `--debug`, I'm not sure how to do that for the dockerized deployment...

**Reply by KarlStraussberger** [reply]
You should have a `srt` entry in `extension-mimetype`. Normally you can open a  ssh connection into your container, stop the service and run gerbera manually

**Reply by Freealx** [reply]
Hi Karl,
Stopping the gerbera process is stopping the all container... however, I managed to change the container CMD parameter in adding '--debug' to it and that works and the debug mode is applied - this is from the container console:
```
/ # ps -ef
PID   USER     TIME  COMMAND
    1 root      0:00 /sbin/tini -- docker-entrypoint.sh gerbera --debug --port 49494 --config /var/run/gerbera/config.xml
    6 gerbera   0:01 gerbera --debug --port 49494 --config /var/run/gerbera/config.xml
   30 root      0:00 ash
   34 root      0:00 ps -ef
/ #
```

In the file `config/config.xml` I have now those SRT related lines:
```
    <mappings>
      <extension-mimetype ignore-unknown="no">
        <map from="...
        <map from="srt" to="text/plain" />            <---------
                   ... />
      </extension-mimetype>
      <mimetype-upnpclass>
        <map from="...
                   ... />
      </mimetype-upnpclass>
      <mimetype-contenttype>
        <treat mimetype="...
                         ... />
      </mimetype-contenttype>
      <mimetype-dlnatransfermode>
        <map from="...
        <map from="srt" to="Background" />            <---------
        <map from="text/*" to="Background" />         <---------
                   ... />
      </mimetype-dlnatransfermode>
      <contenttype-dlnaprofile>
        <map from= ...
                   ... />
      </contenttype-dlnaprofile>
    </mappings>
```

Unfortunately the external subtitles (same name as the video) are still not even available for selection.
The only subs are the embedded in the MKV files. Selecting them is not producing any additional debug lines in the log.

BTW, in the debug mode my Samsung Smart TV cannot even to read the DLNA information from the Gerbera server.
The server log is showing a loop of this output:
```
2022-05-20 23:59:16.606  debug: [upnp_xml.cc:232] renderObject(): Rendered DIDL: <container id="3" parentID="0" restricted="1">
  <dc:title>Video</dc:title>
  <upnp:class>object.container</upnp:class>
  <dc:date>2022-05-20T23:59:16+0000</dc:date>

</container>

2022-05-20 23:59:16.606  debug: [upnp_cds.cc:133] doBrowse(): didl <?xml version="1.0" encoding="UTF-8"?>
<DIDL-Lite xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:sec="http://www.sec.co.kr/dlna">
<container id="3" parentID="0" restricted="1">
<dc:title>Video</dc:title>
<upnp:class>object.container</upnp:class>
<dc:date>2022-05-20T23:59:16+0000</dc:date>

</container>

</DIDL-Lite>

2022-05-20 23:59:16.606  debug: [upnp_cds.cc:143] doBrowse(): end
2022-05-20 23:59:16.607  debug: [upnp_cds.cc:362] processActionRequest(): ContentDirectoryService::processActionRequest: end
2022-05-20 23:59:16.607  debug: [action_request.cc:100] update(): ActionRequest::update(): <u:BrowseResponse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1">
<Result>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;DIDL-Lite xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/" xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:sec="http://www.sec.co.kr/dlna"&gt;
&lt;container id="3" parentID="0" restricted="1"&gt;
&lt;dc:title&gt;Video&lt;/dc:title&gt;
&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;
&lt;dc:date&gt;2022-05-20T23:59:16+0000&lt;/dc:date&gt;
&lt;/container&gt;
&lt;/DIDL-Lite&gt;

>
<NumberReturned>1</NumberReturned>
<TotalMatches>0</TotalMatches>
<UpdateID>0</UpdateID>
</u:BrowseResponse>

2022-05-20 23:59:16.607  debug: [action_request.cc:115] update(): ActionRequest::update(): converted to iXML, code 0
2022-05-20 23:59:16.607  debug: [server.cc:417] handleUpnpRootDeviceEvent(): returning 0
2022-05-20 23:59:16.607  debug: [thread_runner.h:94] wait(): ThreadRunner: Waiting for SQLiteThread
```

Good luck!

**Reply by KarlStraussberger** [reply]
Gerbera does everything on import, so you need to add the above `resources` section and have your files rescanned. After that check out the details on the web ui to make sure we got it imported.

The non-printable char looks dubious.
